core/ArchiveProcessor/RecordBuilder.php

```diff
@@ -0,0 +1,224 @@
+<?php
+/**
+ * Matomo - free/libre analytics platform
+ *
+ * @link    https://matomo.org
+ * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ */
+
+namespace Piwik\ArchiveProcessor;
+
+use Piwik\ArchiveProcessor;
+use Piwik\Common;
+use Piwik\DataTable;
+
+/**
+ * Inherit from this class to define archiving logic for one or more records.
+ *
+ * @since 5.0.0
+ */
+abstract class RecordBuilder
+{
+    /**
+     * @var int
+     */
+    protected $maxRowsInTable;
+
+    /**
+     * @var int
+     */
+    protected $maxRowsInSubtable;
+
+    /**
+     * @var string|int
+     */
+    protected $columnToSortByBeforeTruncation;
+
+    /**
+     * @var int
+     */
+    protected $blobReportLimit;
+
+    /**
+     * @var array|null
+     */
+    protected $columnAggregationOps;
+
+    /**
+     * @param int|null $maxRowsInTable
+     * @param int|null $maxRowsInSubtable
+     * @param string|int|null $columnToSortByBeforeTruncation
+     * @param array|null $columnAggregationOps
+     * @api
+     */
+    public function __construct($maxRowsInTable = null, $maxRowsInSubtable = null,
+                                $columnToSortByBeforeTruncation = null, $columnAggregationOps = null)
+    {
+        $this->maxRowsInTable = $maxRowsInTable;
+        $this->maxRowsInSubtable = $maxRowsInSubtable;
+        $this->columnToSortByBeforeTruncation = $columnToSortByBeforeTruncation;
+        $this->columnAggregationOps = $columnAggregationOps;
+    }
+
+    public function isEnabled()
+    {
+        return true;
+    }
+
+    public function build(ArchiveProcessor $archiveProcessor)
+    {
+        if (!$this->isEnabled()) {
+            return;
+        }
+
+        $numericRecords = [];
+
+        $records = $this->aggregate($archiveProcessor);
+        foreach ($records as $recordName => $recordValue) {
+            if ($recordValue instanceof DataTable) {
+                $this->insertRecord($archiveProcessor, $recordName, $recordValue);
+
+                Common::destroy($recordValue);
+                unset($recordValue);
+            } else {
+                // collect numeric records so we can insert them all at once
+                $numericRecords[$recordName] = $recordValue;
+            }
+        }
+        unset($records);
+
+        if (!empty($numericRecords)) {
+            $archiveProcessor->insertNumericRecords($numericRecords);
+        }
+    }
+
+    public function buildMultiplePeriod(ArchiveProcessor $archiveProcessor)
+    {
+        if (!$this->isEnabled()) {
+            return;
+        }
+
+        $requestedReports = $archiveProcessor->getParams()->getArchiveOnlyReportAsArray();
```

Comment:
the difference between the two is not fully clear yet to me. Is one the requested record someone is fetching and the other one is the records we found?

Reply:
<img width="473" alt="image" src="https://user-images.githubusercontent.com/273120/225189555-0d0b370e-51a7-4ce4-8d39-11e062c34679.png">
actually got it now. It's obvious now that I see it. Once is the reports that were requested, the other one is which ones were actually found. I got a bit confused by `getArchiveOnlyReportAsArray`